### PR TITLE
ARM64-SVE: Add `TrigonometricMultiplyAddCoefficient`

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Part of #99957. @dotnet/arm64-contrib PTAL, thanks!

Test output:
```
Starting test: .\Core_Root\corerun.exe .\HardwareIntrinsics_Arm_r\HardwareIntrinsics_Arm_r.dll Sve_TrigonometricMultiplyAddCoefficient
===================Running default===================
------------------- {} -------------------
Passed test: _Sve_r::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_TrigonometricMultiplyAddCoefficient_float() : 31
Passed test: _Sve_r::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_TrigonometricMultiplyAddCoefficient_double() : 31
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------

Starting test: .\Core_Root\corerun.exe .\HardwareIntrinsics_Arm_ro\HardwareIntrinsics_Arm_ro.dll Sve_TrigonometricMultiplyAddCoefficient
===================Running default===================
------------------- {} -------------------
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_TrigonometricMultiplyAddCoefficient_float() : 31
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_TrigonometricMultiplyAddCoefficient_double() : 31
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------
```

### CODE REVIEW

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp [hunk 1]
```diff
@@ -2297,6 +2297,26 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
...
+                {
+                    assert(targetReg != op2Reg);
+
+                    GetEmitter()->emitInsSve_R_R(INS_sve_movprfx, EA_SCALABLE, targetReg, op1Reg);
+                }
+
+                HWIntrinsicImmOpHelper helper(this, intrin.op3, node);
+
+                for (helper.EmitBegin(); !helper.Done(); helper.EmitCaseEnd())
+                {
```
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2314]: do you mind sharing the codegen for constant and out-of-bound constant/variable input?

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template [hunk 2]
```diff
@@ -150,7 +150,7 @@ namespace JIT.HardwareIntrinsics.Arm
 
             public void RunStructFldScenario({TemplateName}BinaryOpTest__{TestName} testClass)
             {
-                var result = {Isa}.{Method}(_fld1, _fld2, {Imm});
+                var result = {Isa}.{Method}(_fld1, _fld2, Imm);
```
Comment (kunalspathak) [inline src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template:153]: This is missing a test where we pass `invalidImm` as input. See `test.RunBasicScenario_UnsafeRead_InvalidImm` in `_SveImmUnaryOpTestTemplate.template` for example.

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs [hunk 3]
```diff
@@ -5218,6 +5218,33 @@ public static float MultiplyExtended(float op1, float op2)
             }
         }
 
+        public static float TrigonometricMultiplyAddCoefficient(float op1, float op2, byte imm)
+        {
+            int index = (op2 < 0) ? (imm + 8) : imm;
+            uint coeff = index switch
```
Comment (kunalspathak) [inline src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs:5224]: @SwapnilGaikwad - can you please confirm this logic verifies the operation of the instruction?

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs [hunk 4]
```diff
@@ -5218,6 +5218,33 @@ public static float MultiplyExtended(float op1, float op2)
             }
         }
 
+        public static float TrigonometricMultiplyAddCoefficient(float op1, float op2, byte imm)
```
Comment (kunalspathak) [inline src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs:5221]: shouldn't there be `double` equivalent of this helper?

**Review by kunalspathak — CHANGES_REQUESTED** [review]
Added some questions around test.

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs (hunk 4 above)
Comment (amanasifkhalid) [reply src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs:5221]: The `double` version is further down `Helpers.cs`, on line 5292.

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template (hunk 2 above)
Comment (amanasifkhalid) [reply src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template:153]: Got it. I noticed the `InvalidImm` tests in other templates look like this:
```csharp
bool succeeded = false;
try
{
    var result = {Isa}.{Method}(
        Unsafe.Read<{Op1VectorType}<{Op1BaseType}>>(_dataTable.inArrayPtr),
        {InvalidImm}
    );
}
catch (ArgumentOutOfRangeException)
{
    succeeded = true;
}
```

We aren't doing anything with `succeeded`. Is this intentional?

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp (hunk 1 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2314]: Sure. Here's the constant case:
```
movprfx z11, z8
ftmad   z11.d, z11.d, z10.d, #2
```

And out-of-bounds (the intrinsic is gone):
```
G_M52739_IG01:  ;; offset=0x0000
            stp     fp, lr, [sp, #-0x30]!
            stp     x19, x20, [sp, #0x20]
            mov     fp, sp
            add     x1, sp, #48
            str     x1, [fp, #0x18]     // [V14 PSPSym]
            mov     x19, x0
                                                ;; size=24 bbWeight=1 PerfScore 4.50
G_M52739_IG02:  ;; offset=0x0018
            movz    x0, #0x3D00
            movk    x0, #0x7D49 LSL #16
            movk    x0, #686 LSL #32
            movz    x1, #0x5210
            movk    x1, #0x7D49 LSL #16
            movk    x1, #686 LSL #32
            movz    x2, #0xC018      // code for System.String:Concat(System.String,System.String):System.String
            movk    x2, #0x395D LSL #16
            movk    x2, #0x7FFF LSL #32
            ldr     x2, [x2]
            blr     x2
            movz    x1, #0x7DB0      // code for System.Console:WriteLine(System.String)
            movk    x1, #0x39CC LSL #16
            movk    x1, #0x7FFF LSL #32
            ldr     x1, [x1]
            blr     x1
                                                ;; size=64 bbWeight=1 PerfScore 14.00
G_M52739_IG03:  ;; offset=0x0058
            add     x19, x19, #80
            mov     x20, x19
            ldrsb   wzr, [x20]
            add     x0, x20, #32
            movz    x1, #0xCF00      // code for System.Runtime.InteropServices.GCHandle:AddrOfPinnedObject():long:this
            movk    x1, #0x39BF LSL #16
            movk    x1, #0x7FFF LSL #32
            ldr     x1, [x1]
            blr     x1
            ldr     x1, [x20, #0x18]
            add     x0, x0, x1
            sub     x0, x0, #1
            sub     x1, x1, #1
            bic     x0, x0, x1
            ldrsb   wzr, [x0]
            add     x0, x19, #40
            movz    x1, #0xCF00      // code for System.Runtime.InteropServices.GCHandle:AddrOfPinnedObject():long:this
            movk    x1, #0x39BF LSL #16
            movk    x1, #0x7FFF LSL #32
            ldr     x1, [x1]
            blr     x1
            ldr     x1, [x19, #0x18]
            add     x0, x0, x1
            sub     x0, x0, #1
            sub     x1, x1, #1
            bic     x0, x0, x1
            ldrsb   wzr, [x0]
                                                ;; size=108 bbWeight=1 PerfScore 32.00
G_M52739_IG04:  ;; offset=0x00C4
            ldp     x19, x20, [sp, #0x20]
            ldp     fp, lr, [sp], #0x30
            ret     lr
                                                ;; size=12 bbWeight=1 PerfScore 3.00
G_M52739_IG05:  ;; offset=0x00D0
            stp     fp, lr, [sp, #-0x30]!
            stp     x19, x20, [sp, #0x20]
            add     x3, fp, #48
            str     x3, [sp, #0x18]
                                                ;; size=16 bbWeight=0 PerfScore 0.00
G_M52739_IG06:  ;; offset=0x00E0
            adr     x0, [G_M52739_IG04]
                                                ;; size=4 bbWeight=0 PerfScore 0.00
G_M52739_IG07:  ;; offset=0x00E4
            ldp     x19, x20, [sp, #0x20]
            ldp     fp, lr, [sp], #0x30
            ret     lr
                                                ;; size=12 bbWeight=0 PerfScore 0.00
```

And non-const (from the reflection scenario):
```
G_M9075_IG01:  ;; offset=0x0000
            stp     fp, lr, [sp, #-0x10]!
            mov     fp, sp
                                                ;; size=8 bbWeight=1 PerfScore 1.50
G_M9075_IG02:  ;; offset=0x0008
            uxtb    w0, w0
            cmp     w0, #8
            bhs     G_M9075_IG12
            adr     x1, [G_M9075_IG03]
            add     x1, x1, x0,  LSL #3
            br      x1
                                                ;; size=24 bbWeight=1 PerfScore 4.50
G_M9075_IG03:  ;; offset=0x0020
            ftmad   z0.d, z0.d, z1.d, #0
            b       G_M9075_IG11
                                                ;; size=8 bbWeight=1 PerfScore 4.00
G_M9075_IG04:  ;; offset=0x0028
            ftmad   z0.d, z0.d, z1.d, #1
            b       G_M9075_IG11
                                                ;; size=8 bbWeight=1 PerfScore 4.00
G_M9075_IG05:  ;; offset=0x0030
            ftmad   z0.d, z0.d, z1.d, #2
            b       G_M9075_IG11
                                                ;; size=8 bbWeight=1 PerfScore 4.00
G_M9075_IG06:  ;; offset=0x0038
            ftmad   z0.d, z0.d, z1.d, #3
            b       G_M9075_IG11
                                                ;; size=8 bbWeight=1 PerfScore 4.00
G_M9075_IG07:  ;; offset=0x0040
            ftmad   z0.d, z0.d, z1.d, #4
            b       G_M9075_IG11
                                                ;; size=8 bbWeight=1 PerfScore 4.00
G_M9075_IG08:  ;; offset=0x0048
            ftmad   z0.d, z0.d, z1.d, #5
            b       G_M9075_IG11
                                                ;; size=8 bbWeight=1 PerfScore 4.00
G_M9075_IG09:  ;; offset=0x0050
            ftmad   z0.d, z0.d, z1.d, #6
            b       G_M9075_IG11
                                                ;; size=8 bbWeight=1 PerfScore 4.00
G_M9075_IG10:  ;; offset=0x0058
            ftmad   z0.d, z0.d, z1.d, #7
                                                ;; size=4 bbWeight=1 PerfScore 3.00
G_M9075_IG11:  ;; offset=0x005C
            ldp     fp, lr, [sp], #0x10
            ret     lr
                                                ;; size=8 bbWeight=1 PerfScore 2.00
G_M9075_IG12:  ;; offset=0x0064
            bl      CORINFO_HELP_THROW_ARGUMENTOUTOFRANGEEXCEPTION
            brk_windows #0
                                                ;; size=8 bbWeight=0 PerfScore 0.00
```

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template (hunk 2 above)
Comment (kunalspathak) [reply src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template:156]: > We aren't doing anything with `succeeded`. Is this intentional?

Unfortunately, it is not. It is missing this code:

```c#
            if (!succeeded)
            {
                Succeeded = false;
            }
```

Opened https://github.com/dotnet/runtime/issues/104809 to track it. Feel free to just update this one and for others we will take care of as part of fixing the overall issue.

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template [hunk 5]
```diff
@@ -233,7 +255,7 @@ namespace JIT.HardwareIntrinsics.Arm
                                      .Invoke(null, new object[] {
                                         Unsafe.Read<{Op1VectorType}<{Op1BaseType}>>(_dataTable.inArray1Ptr),
                                         Unsafe.Read<{Op1VectorType}<{Op1BaseType}>>(_dataTable.inArray2Ptr),
-                                        (byte){Imm}
+                                        (byte)Imm
```
Comment (kunalspathak) [inline src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template:258]: By having `Imm`, we will always generate table driven logic, while with `{Imm}`, we will always generate the single instruction, with constant embedded. I think we need to have a test for both and I have captured it in https://github.com/dotnet/runtime/issues/104809. Please revert these changes and we will address them together as part of that issue.

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs [hunk 6]
```diff
@@ -5218,6 +5218,33 @@ public static float MultiplyExtended(float op1, float op2)
...
+        {
+            int index = (op2 < 0) ? (imm + 8) : imm;
```
Comment (kunalspathak) [inline src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs:5223]: have we verified the behavior of this API with invalid values? If it is undefined, are we making the test robust to handle it?

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template (hunk 5 above)
Comment (amanasifkhalid) [reply src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template:258]: Got it, I did this out of convenience for writing the test definitions: If we do `{Imm}`, then we cannot dynamically generate the immediate value using something like `TestLibrary.Generator.GetByte()` because the value will change each time we use `{Imm}`. Instead, we'll have to hard-code the immediate into the test with something like `["Imm"] = "0"`. I don't mind doing this; it will just result in more tests for this API.

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs (hunk 6 above)
Comment (amanasifkhalid) [reply src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs:5223]: Like in #104681, the tests seem to work fine with out-of-bound values, though I can easily modify the tests to skip validation for such values.

**Comment by amanasifkhalid** [comment]
@kunalspathak I addressed your feedback. I had to add a usage of the intrinsic's result in the `InvalidImm` scenario so that the intrinsic itself doesn't get optimized away and fail the test. Stress tests are passing.

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template [hunk 7]
```diff
@@ -62,6 +62,9 @@ namespace JIT.HardwareIntrinsics.Arm
 
                 // Validates executing the test inside conditional, with op3 as zero
                 test.ConditionalSelect_ZeroOp();
+
+                // Validates basic functionality fails with an invalid imm, using Unsafe.ReadUnaligned
+                test.RunBasicScenario_UnsafeRead_InvalidImm();
```
Comment (kunalspathak) [inline src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveImmBinaryOpTestTemplate.template:67]: This template is used by `*MultiplyBySelectedScalar*` too. Have you verified if those tests pass, including the stress?

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs (hunk 6 above)
Comment (kunalspathak) [reply src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs:5223]: I liked the way you have combined the table of `sin` and `cos` in a single lookup.

Comment (kunalspathak) [reply src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs:5223]: > Like in #104681, the tests seem to work fine with out-of-bound values, though I can easily modify the tests to skip validation for such values.

That will be great.

**Review by kunalspathak — CHANGES_REQUESTED** [review]
Overall looks good. please make sure we skip the validation for invalid input as well as make sure the other tests using the modified template passes.

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs (hunk 6 above)
Comment (amanasifkhalid) [reply src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs:5223]: > I liked the way you have combined the table of sin and cos in a single lookup.

Thanks, I wish I could say that was my idea, but the ARM docs suggested it

**Comment by amanasifkhalid** [comment]
@kunalspathak I merged from main and reran the stress tests for `MultiplyBySelectedScalar` and `TrigonometricMultiplyAddCoefficient`, and they're passing. Is this ready to be merged?

**Review by kunalspathak — APPROVED** [review]
LGTM. Thanks!
